collector/pg_replication_slots.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func init() {
+	registerCollector("replication_slot", defaultEnabled, NewPGReplicationSlotCollector)
+}
+
+type PGReplicationSlotCollector struct {
+	log log.Logger
+}
+
+func NewPGReplicationSlotCollector(logger log.Logger) (Collector, error) {
+	return &PGReplicationSlotCollector{log: logger}, nil
+}
+
+var pgReplicationSlot = map[string]*prometheus.Desc{
+	"current_wal_lsn": prometheus.NewDesc(
+		"pg_replication_slot_current_wal_lsn",
+		"current wal lsn value",
+		[]string{"slot_name"}, nil,
+	),
+	"confirmed_flush_lsn": prometheus.NewDesc(
+		"pg_replication_slot_confirmed_flush_lsn",
+		"last lsn confirmed flushed to the replication slot",
+		[]string{"slot_name"}, nil,
+	),
+}
+
+func (PGReplicationSlotCollector) Update(ctx context.Context, db *sql.DB, ch chan<- prometheus.Metric) error {
+	rows, err := db.QueryContext(ctx,
+		`SELECT
+			slot_name,
+			pg_current_wal_lsn() AS current_wal_lsn,
+			confirmed_flush_lsn
+			FROM
+		pg_replication_slots;`)
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var slot_name string
+		var wal_lsn int64
+		var flush_lsn int64
```

Comment:
The docs say that this could be null so this should be a sql.NullInt64 to account for that. I'm not sure what the value should be in that case. I suppose 0 would be fine. 

https://www.postgresql.org/docs/current/view-pg-replication-slots.html

Reply:
i agree with you that it should be 0, i'll also add a bool counter for is_active to denote whether or not the 0 is actually useful here.

Reply:
Usually for null, we don't expose the metric.

"is_active" should probably be a 0/1 boolean gauge like `pg_replication_lsn_active` or something like that.